feat: CWで注釈を空欄のままノートできないようにする

sanao1006 opened the pull request:
## 概要
misskeyのバージョンが11.1以上の時、「内容を隠す」で注釈が空の時にノートできなくなった https://github.com/misskey-dev/misskey/issues/12217
上記条件を満たすサーバーにmilkteaでログインして注釈が空のままノートするとエラーが発生する

## やったこと
milkteaのバージョンが　"2023.11.0"　以上の時に注釈が空欄のままノートできないようにした
## 動作確認
https://github.com/pantasystem/Milktea/assets/62137820/5a78ed47-3d80-4f00-bd0c-a5fb9e09ebf2


## 懸念事項
- coroutineScopeに `Dispatchers.IO` を加えるべきかが分からない
    - とりあえず IO が絡んでないように見えるのでつけなかった
- 今回の `started` は自分の感覚では `WhileSubscribed` なのだが`Eagerly`にしないと値が取れなかった
    - CWのボタンを押して注釈欄が出てきたときのみアクティブになればよいと思ったのですが...
- そもそも新しく変数 `private val isCwAllowBlank` を作らなくても既存変数を使えばうまく実装できたかも？

## Issue番号

Fixed #1992

sanao1006 committed:
feat: CWで注釈を空欄のままノートできないようにする

misskeyのバージョンが11.1以上の時、「内容を隠す」で注釈が空の時にノートできなくなったため

Fixed #1992

sanao1006 commented:
気づいたけどこの実装では文字が入っていようがそもそもノートができない

sanao1006 commented:
misskey バージョンが "2023.11.0" 以上ならば、テキストの入力にかかわらず問答無用で `checkValidate` が false を返すような実装になっているようなきがする

pantasystem commented on modules/features/note/src/main/java/net/pantasystem/milktea/note/editor/viewmodel/NoteEditorViewModel.kt:
```diff
@@ -242,7 +246,7 @@ class NoteEditorViewModel @Inject constructor(
     }.stateIn(viewModelScope + Dispatchers.IO, started = SharingStarted.Lazily, initialValue = 1500)
 
     val isPostAvailable = uiState.map {
```
この際combineに置き換えれば良いのではないかと思いました。
https://github.com/pantasystem/Milktea/pull/1994/files#diff-1c9bf7e0caaa1181bc7f1af78e457828afad596a7841266e832cb9d9e2441094R126-R129 がWhileSubscribedで動作しなかった理由としては
どこからも購読されていなかったため、Flowが動作しませんでした。
なので期待値通り動作させたければcombineなどから使用するようにして、
isPostAvailable購読時に一緒に購読が開始されるようにしないといけないです。

pantasystem commented on modules/features/note/src/main/java/net/pantasystem/milktea/note/editor/viewmodel/NoteEditorUiState.kt:
```diff
@@ -58,11 +58,15 @@ data class NoteEditorUiState(
             return false
         }
 
+        if(!isCwAllowBlank) {
```
おっしゃる通り、isCwAllowBlankはあくまでもCW欄が空欄を許容するのか？ということを表すフィールドなので、isCwAllowBlankがfalseかつcwがblankの時はfalseを返すという処理にする必要性があると思います。

pantasystem reviewed (commented):
まだDraftですが失礼します

sanao1006 committed:
fix: 文字を入力していてもノートができないバグの修正

misskeyのバージョンチェックしかしておらず、CWが空かどうかの判定をしていなかった。
これにより発生していた、テキストを入力していても文字が入力できないバグを修正する

Fixed #1992

sanao1006 committed:
fix: isCwAllowBlankをcombine内にまとめた

以前までのコードでは、`isCwAllowBlank`は誰からも購読されておらずアクティブになっていなかったため無理やりEagerlyを使うことで値を取得していた。
これはリソース上効率が悪いためcombineを用いて、isPostAvailableが購読された時に一緒に購読が開始されるように修正する。

Fixed #1992

sanao1006 commented:
文字を入れるとノートできるようになった
<video src="https://github.com/pantasystem/Milktea/assets/62137820/3c5f47bf-62f0-4721-adb0-232e77ae1c3f"  />

pantasystem reviewed (approved):
良いと思います！🙇
![LGTM](https://image.lgtmoon.dev/229680)
